File: Problem_10.java
class Solution {

    /*
     * Time Complexity - O(logn)
     * Space Complexity - O(1)
     */
    public int findPeakElement(int[] nums) {
        if (nums.length == 1) return 0;
        int low = 0, high = nums.length-1;
        while (low <= high) {
            int mid = low + (high-low)/2;
            if ((mid == 0 || nums[mid] > nums[mid-1]) 
                && (mid == nums.length-1 || nums[mid] > nums[mid+1]))
                return mid;
            else if (mid > 0 && nums[mid] < nums[mid-1])
                high = mid-1;
            else 
                low = mid + 1;
        }
        return -1;
    }
}

File: Problem_10.py
# Time Complexity: O(log n)
# Space Complexity: O(1)

class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums) - 1

        if nums[-1] >= nums[0]:
            return nums[0]

        while left <= right:
            mid = left + (right - left // 2 )

            if nums[left] < nums[mid]:
                left = mid + 1
            elif nums[left] > nums[mid]:
                if nums[mid - 1] > nums[mid]:
                    return nums[mid]
                else:
                    right = mid - 1
        
        return -1